librad/src/git/types/reference.rs

```diff
@@ -20,24 +20,23 @@ use std::{
     marker::PhantomData,
 };
 
-use crate::{git::ext, hash::Hash, uri::RadUrn};
+use crate::{git::ext, uri::RadUrn};
 
-use super::{
-    existential::{SomeNamespace, SomeReference},
-    Force,
-    Refspec,
-    SymbolicRef,
-};
+use super::{namespace::AsNamespace, Force, Refspec, SymbolicRef};
 
 /// Type witness for a [`Reference`] that should point to a single reference.
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-pub struct Single;
+pub struct One;
+
+/// Alias for [`One`].
+pub type Single = One;
```

Comment:
Do you think we need this or are you just planning on doing this change over in phases?

Reply:
It's the Rust version of unison :grimacing:

Reply:
lol, well can we stick with `One`? Makes more sense tbh

Reply:
Actually, the thing is that `name` is not well-defined -- what we want is either a `ext::RefLike` (if it is not a pattern), or a corresponding type which passes the validation rules with `GIT_REFERENCE_FORMAT_PATTERN`. So, the cardinality parameter would actually become obsolete (since we can't have the type of `name` _depend_ on it).

Then we could also define meaningful traits for things which can be in a refspec or symref -- `AsRefspec` boils down to just `Display` on both sides (which was the case before this patch). What we want is to restrict `Refspec` to `Local: Into<Either<RefLike, RefPattern>>, Remote: Into<Either<RefLike, RefPattern>>` and `SymbolicRef` to `S: Into<RefLike>, T: Into<RefLike>`.

Reply:
Ya, that makes sense :ok_hand: Shall we capture that in a follow up?